File: backend/app/schemas/ai.py
from datetime import date, datetime
from typing import List, Literal, Optional, Union

from pydantic import BaseModel, ConfigDict, Field, StrictBool, StrictInt, StrictStr, field_validator, model_validator

PROMPT_VERSION = "v2.0"
CONTEXT_SCHEMA_VERSION = "v2.0"
OUTPUT_SCHEMA_VERSION = "v2.0"
BACKTEST_CONTEXT_VERSION = "v3.backtest.1"
BACKTEST_NEWS_DISCLOSURE = "本报告未纳入新闻数据"
AnalysisMode = Literal["standard", "custom_backtest"]


class StrictSchema(BaseModel):
    model_config = ConfigDict(extra="forbid", protected_namespaces=(), allow_inf_nan=False)


class AIAnalyzeRequest(StrictSchema):
    stock_code: str = Field(pattern=r"^\d{6}$")
    backtest_id: Optional[StrictInt] = Field(
        default=None, gt=0, le=9_223_372_036_854_775_807
    )

    @field_validator("backtest_id", mode="before")
    @classmethod
    def reject_explicit_null(cls, value):
        if value is None:
            raise ValueError("backtest_id must be omitted or a positive integer")
        return value


class StockAnalysisContext(StrictSchema):
    stock_code: str = Field(pattern=r"^\d{6}$")
    stock_name: str = Field(min_length=1)
    industry: Optional[str] = None


class MarketSnapshotContext(StrictSchema):
    trade_date: date
    close: Optional[float] = None
    change_pct: Optional[float] = None
    turnover_rate: Optional[float] = None


class TechnicalIndicatorContext(StrictSchema):
    trade_date: date
    ma5: Optional[float] = None
    ma10: Optional[float] = None
    ma20: Optional[float] = None
    ma60: Optional[float] = None
    macd: Optional[float] = None
    macd_signal: Optional[float] = None
    macd_hist: Optional[float] = None
    rsi14: Optional[float] = None
    boll_upper: Optional[float] = None
    boll_middle: Optional[float] = None
    boll_lower: Optional[float] = None


class QuantScoreContext(StrictSchema):
    score: int = Field(ge=0, le=100)
    level: str = Field(min_length=1)
    reasons: List[str] = Field(default_factory=list)


class BacktestMetricsContext(StrictSchema):
    strategy_name: str = Field(min_length=1)
    start_date: date
    end_date: date
    total_return: Optional[float] = None
    annual_return: Optional[float] = None
    max_drawdown: Optional[float] = None
    sharpe_ratio: Optional[float] = None
    win_rate: Optional[float] = None
    trade_count: Optional[int] = Field(default=None, ge=0)
    benchmark_return: Optional[float] = None


class NewsItemContext(StrictSchema):
    title: str = Field(min_length=1)
    summary: Optional[str] = None
    source: Optional[str] = None
    publish_time: Optional[datetime] = None
    url: Optional[str] = None


SourceMode = Literal["live", "cache", "frozen", "unknown"]
SnapshotStatus = Literal["complete", "legacy_missing"]


class MarketDataProvenance(StrictSchema):
    source_mode: SourceMode
    provider: str = Field(min_length=1)
    market_start_date: date
    market_end_date: date
    market_rows: int = Field(ge=1)


class DataProvenance(MarketDataProvenance):
    news_status: Literal["available", "empty"]
    news_count: int = Field(ge=0)
    retrieved_at: datetime


class AnalysisContext(StrictSchema):
    stock: StockAnalysisContext
    market_snapshot: MarketSnapshotContext
    technical_indicators: Optional[TechnicalIndicatorContext] = None
    quant_score: Optional[QuantScoreContext] = None
    backtest_metrics: Optional[BacktestMetricsContext] = None
    news: List[NewsItemContext] = Field(default_factory=list)
    data_as_of: datetime
    provenance: DataProvenance


class BacktestParametersContext(StrictSchema):
    ma_short_period: StrictInt = Field(ge=2, le=120)
    ma_long_period: StrictInt = Field(ge=2, le=120)
    initial_cash: float = Field(gt=0)
    transaction_cost: float = Field(ge=0, lt=1)
    slippage: float = Field(ge=0, lt=1)

    @model_validator(mode="after")
    def ordered_periods(self):
        if self.ma_short_period >= self.ma_long_period:
            raise ValueError("short MA must be less than long MA")
        return self


class BacktestWarmupContext(StrictSchema):
    start_date: date
    end_date: date
    required_rows: StrictInt = Field(gt=0)
    used_rows: StrictInt = Field(gt=0)


class BacktestInitialEquityContext(StrictSchema):
    trade_date: date
    equity: float = Field(gt=0)
    valuation: Literal["before_open"]


class SavedBacktestMetricsContext(StrictSchema):
    initial_cash: float = Field(gt=0)
    final_equity: float = Field(ge=0)
    total_return: float
    annual_return: float
    max_drawdown: float
    sharpe_ratio: Optional[float]
    win_rate: Optional[float] = Field(ge=0, le=1)
    trade_count: StrictInt = Field(ge=0)
    order_count: StrictInt = Field(ge=0)
    benchmark_return: float


class BacktestInterpretationContext(StrictSchema):
    stock_code: str = Field(pattern=r"^\d{6}$")
    backtest_id: StrictInt = Field(gt=0)
    backtest_created_at: datetime
    data_as_of: datetime
    strategy_name: Literal["ma_long_only"]
    semantics_version: Literal["v2_windowed"]
    algorithm_version: Literal["ma_long_only_v2.0.0"]
    effective_parameters: BacktestParametersContext
    requested_start_date: date
    requested_end_date: date
    start_date: date
    end_date: date
    warmup: BacktestWarmupContext
    execution_assumptions: dict[str, Union[StrictStr, StrictBool]]
    metrics: SavedBacktestMetricsContext
    initial_equity: BacktestInitialEquityContext
    data_hash: str = Field(pattern=r"^[0-9a-f]{64}$")
    provenance: MarketDataProvenance

    @model_validator(mode="after")
    def consistent_saved_context(self):
        if not (self.requested_start_date <= self.start_date <= self.end_date <= self.requested_end_date):
            raise ValueError("inconsistent saved backtest window")
        if not (self.warmup.start_date <= self.warmup.end_date < self.requested_start_date):
            raise ValueError("inconsistent saved warmup window")
        if self.warmup.required_rows != self.effective_parameters.ma_long_period or self.warmup.used_rows != self.warmup.required_rows:
            raise ValueError("inconsistent warmup row count")
        if self.data_as_of != self.backtest_created_at:
            raise ValueError("data_as_of must identify the saved backtest")
        if self.initial_equity.trade_date != self.start_date or self.initial_equity.equity != self.metrics.initial_cash or self.metrics.initial_cash != self.effective_parameters.initial_cash:
            raise ValueError("inconsistent initial equity")
        return self


ReportContext = Union[AnalysisContext, BacktestInterpretationContext]


class AIAnalysisStructuredOutput(StrictSchema):
    trend: Literal["bullish", "neutral", "bearish"]
    summary: str = Field(min_length=1)
    technical_analysis: str = Field(min_length=1)
    quant_analysis: str = Field(min_length=1)
    news_analysis: str = Field(min_length=1)
    advantages: List[str]
    risks: List[str]
    conclusion: str = Field(min_length=1)

    @field_validator(
        "summary",
        "technical_analysis",
        "quant_analysis",
        "news_analysis",
        "conclusion",
    )
    @classmethod
    def reject_blank_text(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("analysis text must not be blank")
        return normalized

    @field_validator("advantages", "risks")
    @classmethod
    def normalize_text_items(cls, value: List[str]) -> List[str]:
        normalized = [item.strip() for item in value if item.strip()]
        if not normalized:
            raise ValueError("analysis list must contain at least one non-blank item")
        return normalized


class AIAnalysisData(AIAnalysisStructuredOutput):
    stock_code: str = Field(pattern=r"^\d{6}$")
    quant_score: Optional[int] = Field(default=None, ge=0, le=100)
    model_name: str = Field(min_length=1)


class AIReportMetadata(StrictSchema):
    analysis_mode: AnalysisMode = "standard"
    backtest_id: Optional[StrictInt] = Field(default=None, gt=0)
    data_as_of: datetime
    source_mode: SourceMode
    prompt_version: str = Field(min_length=1, max_length=32)
    context_schema_version: str = Field(min_length=1, max_length=32)
    output_schema_version: str = Field(min_length=1, max_length=32)
    context_hash: str = Field(pattern=r"^[0-9a-fA-F]{64}$")


class AIReportSummary(StrictSchema):
    analysis_mode: AnalysisMode = "standard"
    backtest_id: Optional[StrictInt] = Field(default=None, gt=0)
    report_id: int = Field(gt=0)
    stock_code: str = Field(pattern=r"^\d{6}$")
    quant_score: Optional[int] = Field(default=None, ge=0, le=100)
    trend: Literal["bullish", "neutral", "bearish"]
    summary: str = Field(min_length=1)
    model_name: str = Field(min_length=1)
    data_as_of: Optional[datetime] = None
    created_at: datetime
    source_mode: SourceMode
    snapshot_status: SnapshotStatus


class AIReportDetail(AIAnalysisData):
    analysis_mode: AnalysisMode = "standard"
    backtest_id: Optional[StrictInt] = Field(default=None, gt=0)
    report_id: int = Field(gt=0)
    created_at: datetime
    data_as_of: Optional[datetime] = None
    source_mode: SourceMode
    prompt_version: Optional[str] = None
    context_schema_version: Optional[str] = None
    output_schema_version: Optional[str] = None
    context_hash: Optional[str] = Field(default=None, pattern=r"^[0-9a-fA-F]{64}$")
    snapshot_status: SnapshotStatus
    context_snapshot: Optional[ReportContext] = None


class PaginatedAIReports(StrictSchema):
    items: List[AIReportSummary]
    total: int = Field(ge=0)
    page: int = Field(ge=1)
    page_size: int = Field(ge=1, le=100)

File: frontend/src/mocks/ai.ts
// AI 分析 mock：字段对齐 docs/API_SPEC.md 第 9 节
import type {
  AIAnalysisData,
  AIReportDetail,
  AIReportSummary,
  ApiResponse,
  PaginatedAIReports,
} from '../types/api'

export function mockAIAnalysis(stockCode: string): ApiResponse<AIAnalysisData> {
  return {
    code: 0,
    message: 'success',
    data: {
      stock_code: stockCode,
      quant_score: 72,
      trend: 'bullish',
      summary: `综合来看，${stockCode} 当前处于短期趋势偏强、中期震荡上行阶段。技术面动能延续，量化评分中等偏上，消息面无明显利空。整体维持谨慎乐观判断。`,
      technical_analysis:
        '价格站上 MA20 且 MA5 上穿 MA10，短期均线呈多头排列；MACD 柱状图在零轴上方温和放大；RSI14 约 62，处于强势区间但尚未超买。布林带上轨附近有压力，需关注放量突破情况。',
      quant_analysis:
        '量化综合评分 72 分（满分 100），其中趋势分 30/40、动量分 18/25、量能分 12/20、风险分 12/15。趋势与动量贡献主要得分，量能略有不足。',
      news_analysis:
        '近期无重大负面公告，行业景气度中性偏暖，市场情绪稳定。未检测到可能引发剧烈波动的突发事件。',
      advantages: [
        '短期均线多头排列，趋势结构清晰',
        'RSI 强势区间运行且未超买，上行仍有空间',
        '回撤控制良好，风险分项得分较高',
      ],
      risks: [
        '接近布林带上轨，存在技术性回调压力',
        '量能未能同步放大，突破需要成交量确认',
        '估值处于历史偏高分位，注意情绪波动',
      ],
      conclusion:
        '建议以中性偏多思路对待：持仓者可继续持有并关注 MA20 支撑，空仓者等待回调至均线附近再考虑介入，突破布林带上轨且放量时可视为趋势加强信号。',
      model_name: 'mock-model',
    },
  }
}

// ============ V2 AI 报告历史 mock（契约：D 的 V2_AI_REPORT_HISTORY_CONTRACT.md）============
// 覆盖两种记录形态：id=101 完整快照（complete）+ id=100 旧记录（legacy_missing）

const MOCK_REPORT_SUMMARIES: AIReportSummary[] = [
  {
    report_id: 101,
    stock_code: '600519',
    quant_score: 82,
    trend: 'bullish',
    summary:
      '趋势得分较高，均线多头排列延续，量能配合尚可，综合评分处于强势区间，维持谨慎乐观。',
    model_name: 'deepseek-v4-flash',
    data_as_of: '2026-09-14T15:00:00Z',
    created_at: '2026-09-15T08:30:00Z',
    source_mode: 'live',
    snapshot_status: 'complete',
  },
  {
    report_id: 100,
    stock_code: '600519',
    quant_score: 33,
    trend: 'bearish',
    summary: '早期冻结演示报告：趋势偏弱，成交低迷，评分处于较弱区间，建议观望。',
    model_name: 'deepseek-v4-flash',
    data_as_of: null,
    created_at: '2026-09-10T02:12:00Z',
    source_mode: 'unknown',
    snapshot_status: 'legacy_missing',
  },
]

export function mockAIReports(params: {
  stock_code?: string
  page?: number
  page_size?: number
}): ApiResponse<PaginatedAIReports> {
  const page = params.page ?? 1
  const pageSize = params.page_size ?? 20
  const filtered = params.stock_code
    ? MOCK_REPORT_SUMMARIES.filter((item) => item.stock_code === params.stock_code)
    : MOCK_REPORT_SUMMARIES
  const start = (page - 1) * pageSize
  return {
    code: 0,
    message: 'success',
    data: {
      items: filtered.slice(start, start + pageSize),
      total: filtered.length,
      page,
      page_size: pageSize,
    },
  }
}

export function mockAIReportDetail(reportId: number): ApiResponse<AIReportDetail> {
  const base = mockAIAnalysis('600519').data
  if (reportId === 100) {
    // V1 旧记录：无上下文快照、无版本字段，来源未知
    return {
      code: 0,
      message: 'success',
      data: {
        ...base,
        quant_score: 33,
        trend: 'bearish',
        report_id: 100,
        created_at: '2026-09-10T02:12:00Z',
        data_as_of: null,
        source_mode: 'unknown',
        prompt_version: null,
        context_schema_version: null,
        output_schema_version: null,
        context_hash: null,
        snapshot_status: 'legacy_missing',
        context_snapshot: null,
      },
    }
  }
  return {
    code: 0,
    message: 'success',
    data: {
      ...base,
      quant_score: 82,
      report_id: reportId,
      created_at: '2026-09-15T08:30:00Z',
      data_as_of: '2026-09-14T15:00:00Z',
      source_mode: 'live',
      prompt_version: 'v2.0',
      context_schema_version: 'v2.0',
      output_schema_version: 'v2.0',
      context_hash:
        '9af45d325e7cced7fc420da9f502e3abcf72ede1b7ad95070b264f8b0c4052f4',
      snapshot_status: 'complete',
      context_snapshot: {
        stock: { stock_code: '600519', stock_name: '贵州茅台', industry: '白酒' },
        market_snapshot: {
          trade_date: '2026-09-14',
          close: 1450.5,
          change_pct: 0.012,
          turnover_rate: 0.002,
        },
        technical_indicators: null,
        quant_score: { score: 82, level: 'strong', reasons: ['趋势得分较高'] },
        backtest_metrics: null,
        news: [],
        data_as_of: '2026-09-14T15:00:00Z',
        provenance: {
          source_mode: 'live',
          provider: 'akshare',
          market_start_date: '2025-09-14',
          market_end_date: '2026-09-14',
          market_rows: 243,
          news_status: 'empty',
          news_count: 0,
          retrieved_at: '2026-09-14T15:00:00Z',
        },
      },
    },
  }
}
